File: Demonstration/matlab/AudioNoiseGate.m
classdef AudioNoiseGate
    %AUDIONOISEGATE Creates an audio noise gate
    % class which can be used to perform a noise gate effect on audio
    
    properties
        threshold           % Threshold parameter (in decibels)
        thresholdLin        % Threshold parameter (in magnitude)
        attack              % Attack time parameter
        release             % Release time parameter
        hold                % Hold time parameter
    end
    
    methods
        function obj = AudioNoiseGate(threshold, hold, attack, release)
            % Constructor to initialize the object with specified parameters
            %
            % Parameters
            % threshold: the dB threshold value that the compressor will
            %   start to take effect after passing
            % hold: the amount of time to hold the compression on after the
            %   attack phase and before the release phase
            % attack: the ammount of time it takes for the compression to
            %   start taking effect after passing the threshold
            % release: the ammount of time it takes for the compression to
            %   stop being applied after going above the threshold

            obj.threshold = threshold;
            obj.thresholdLin = decibelsToGain(threshold, -120);
            obj.attack = attack;
            obj.release = release;
            obj.hold = hold;
        end
        
        function [y, gs] = process(obj, x, fs, sidechain)
            % Public method to process the input signal
            % The entire processing chain is applied here
            %
            % Parameters
            % x: input signal
            % fs: sample rate of the input signal
            % sidechain (OPTIONAL): the sidechain input to use to calculate
            % gain reduction
            
            if ~exist('sidechain','var')
                 % If sidechain doesn't exist, use input as sidechain
                  sidechain = x;
            end

            % Calculate the absolute of the N-point signal (in magnitude)
            xa = abs(sidechain);
            
            % Compute the static characteristic with soft knee
            gclin = obj.applyLinearStaticCharacteristic(xa);
            
            % Smooth the gain using specified attack and release time parameters
            alphaA = obj.calculateTimeCoeff(obj.attack, fs);
            alphaR = obj.calculateTimeCoeff(obj.release, fs);
            
            glin = obj.smoothGain(gclin, alphaA, alphaR);
            
            % Returning the dB values of the gain reduction
            gs = gainToDecibels(glin, -120);
            
            % Apply the dynamic range limiter
            y = x .* glin;
        end
    end
    
    methods
        function gclin = applyLinearStaticCharacteristic(obj, xa)
            % Public method to compute gain using static characteristic
            % Used for linear calculation
            
            % Either 0 or 1 if over the threshold or not
            gclin = zeros(size(xa));
            gclin(xa >= obj.thresholdLin) = 1;
        end

        function gc = applydBStaticCharacteristic(obj, xa)
            % Public method to compute gain using static characteristic
            % Used for decibel calculation
            
            % Either 0 or 1 if over the threshold or not
            gc = zeros(size(xa));
            gc(xa < obj.threshold) = -60;
        end
        
        function glin = smoothGain(obj, gc, alphaA, alphaR)
            % Public method to apply smoothing through interpolation 
            % onto the gain computer using the attack and release
            % coefficients

            glin = zeros(size(gc));
            
            % Initialize the first element of gs
            glin(1) = gc(1);
            
            % Hold counter for attack
            CA = 0;

            % Perform smoothing using specified attack and release time parameters
            for n = 2:length(gc)
                if (CA > obj.hold) && (gc(n) <= glin(n-1))
                    % Attack phase
                    glin(n) = alphaA * glin(n-1) + (1 - alphaA) * gc(n);

                elseif (CA <= obj.hold)
                    % Hold phase
                    glin(n) = glin(n-1);
                    CA = CA + 1;

                elseif gc(n) > glin(n-1)
                    % Release phase
                    glin(n) = alphaR * glin(n-1) + (1 - alphaR) * gc(n);
                end
            end
        end
        
        function alpha = calculateTimeCoeff(~, time, fs)
            % Public method to calculate the time coefficients for both
            % attack and realease in an exponential form
            alpha = exp(-log(9) / (fs * time));
        end
    end
end

File: Demonstration/PythonCode/AudioNoiseGate.py
import numpy as np
from PythonCode.decibelsToGain import decibels_to_gain
from PythonCode.gainToDecibels import gain_to_decibels

class AudioNoiseGate:
    ''' Creates an audio noise gate class which can be used to perform a noise gate effect on audio. '''
    
    def __init__(self, threshold, hold, attack, release):
        '''
          Constructor to initialize the object with specified parameters.
          
          Parameters -
          threshold: the dB threshold value that the compressor will
            start to take effect after passing
          hold: the amount of time to hold the compression on after the
            attack phase and before the release phase
          attack: the amount of time it takes for the compression to
            start taking effect after passing the threshold
          release: the amount of time it takes for the compression to
            stop being applied after going above the threshold
        '''
        
        self.threshold = threshold
        self.thresholdLin = decibels_to_gain(threshold, -120)
        self.attack = attack
        self.release = release
        self.hold = hold

    def process(self, x, fs, sidechain=None):
        '''
          Public method to process the input signal. The entire processing chain is applied here.
          
          Parameters -
          x: input signal
          fs: sample rate of the input signal
          sidechain (OPTIONAL): the sidechain input to use to calculate
            gain reduction
        '''
        if sidechain is None:
            sidechain = x.copy()

        # Calculate the absolute of the N-point signal (in magnitude)
        xa = np.abs(sidechain)

        # Compute the static characteristic with a soft knee
        gclin = self.applyLinearStaticCharacteristic(xa)

        # Smooth the gain using specified attack and release time parameters
        alphaA = self.calculateTimeCoeff(self.attack, fs)
        alphaR = self.calculateTimeCoeff(self.release, fs)
        glin = self.smoothGain(gclin, alphaA, alphaR)

        # Returning the dB values of the gain reduction
        gs = gain_to_decibels(glin, -120)

        # Apply the dynamic range limiter
        y = x * glin

        return y, gs

    def applyLinearStaticCharacteristic(self, xa):
        ''' Public method to compute gain using static characteristic. Used for linear calculation. '''
        # Either 0 or 1 if over the threshold or not
        gclin = np.zeros_like(xa)
        gclin[xa >= self.thresholdLin] = 1
        return gclin

    def applydBStaticCharacteristic(self, xa):
        ''' Public method to compute gain using static characteristic. Used for decibel calculation. '''
        # Either 0 or 1 if over the threshold or not
        gc = np.zeros_like(xa)
        gc[xa < self.threshold] = -60
        return gc

    def smoothGain(self, gc, alphaA, alphaR):
        '''
          Public method to apply smoothing through interpolation onto the gain computer
          using the attack and release coefficients.
        '''
        glin = np.zeros_like(gc)

        # Initialize the first element of gs
        glin[0] = gc[0]

        # Hold counter for attack
        CA = 0

        # Perform smoothing using specified attack and release time parameters
        for n in range(1, len(gc)):
            if (CA > self.hold) and (gc[n] <= glin[n-1]):
                # Attack phase
                glin[n] = alphaA * glin[n-1] + (1 - alphaA) * gc[n]
            elif (CA <= self.hold):
                # Hold phase
                glin[n] = glin[n-1]
                CA += 1
            elif gc[n] > glin[n-1]:
                # Release phase
                glin[n] = alphaR * glin[n-1] + (1 - alphaR) * gc[n]

        return glin

    def calculateTimeCoeff(self, time, fs):
        '''
          Public method to calculate the time coefficients for both attack and release in an
          exponential form.
        '''
        return np.exp(-np.log(9) / (fs * time))
